### PR TITLE
Add quota config parser

### PR DESCRIPTION
**Pull request by qiwzhang**

### CODE REVIEW

File: src/envoy/mixer/quota_config.cc [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    const std::string& value = it->second.str_v;
+
+    switch (match.match_type_case()) {
+      case ::istio::proxy::v1::config::StringMatch::kExact:
+        if (value != match.exact()) {
+          return false;
+        }
+        break;
+      case ::istio::proxy::v1::config::StringMatch::kPrefix:
+        if (value.length() < match.prefix().length() ||
```
Comment (lizan) [inline src/envoy/mixer/quota_config.cc]: nit: use `StringUtil::startsWith`?

File: src/envoy/mixer/quota_config.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+// attributes to generate required quotas.
+class QuotaConfig {
+ public:
+  QuotaConfig(const ::istio::mixer::v1::config::client::QuotaSpec& spec_pb);
+
+  struct Quota {
+    std::string quota;
+    int64_t charge;
+
+    // For unit-test.
```
Comment (lizan) [inline src/envoy/mixer/quota_config.h]: If this is for unit test only, just add a matcher for gtest.

File: src/envoy/mixer/quota_config.cc (hunk 1 above)
Comment (qiwzhang) [reply src/envoy/mixer/quota_config.cc]: STL doesn't have StringUtil::startsWith

File: src/envoy/mixer/quota_config.h (hunk 2 above)
Comment (qiwzhang) [reply src/envoy/mixer/quota_config.h]: It is used in 
ASSERT_EQ for std::vector<Quota>
gTest matcher will not help.

Comment (qiwzhang) [reply src/envoy/mixer/quota_config.h]: std::vector of Quota

File: src/envoy/mixer/quota_config_test.cc [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+)";
+
+using QuotaVector = std::vector<QuotaConfig::Quota>;
+
+class QuotaConfigTest : public ::testing::Test {
+ public:
+  void SetUp() {}
+};
+
+TEST_F(QuotaConfigTest, TestNotMatch) {
```
Comment (ayj) [inline src/envoy/mixer/quota_config_test.cc]: nit: `TestNotMatch` suggests this spec doesn't match anything. Maybe rename `NotMatch` to `EmptyMatch` here and line 30 to make the intent clearer?

File: src/envoy/mixer/quota_config.cc (hunk 1 above)
Comment (lizan) [reply src/envoy/mixer/quota_config.cc]: Envoy has...

Comment (qiwzhang) [reply src/envoy/mixer/quota_config.cc]: Try not to use Envoy code in this module.  May move it inside MixerClient library.

File: src/envoy/mixer/quota_config_test.cc (hunk 3 above)
Comment (qiwzhang) [reply src/envoy/mixer/quota_config_test.cc]: Done

File: src/envoy/mixer/quota_config.cc [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            value.compare(0, match.prefix().length(), match.prefix()) != 0) {
+          return false;
+        }
+        break;
+      case ::istio::proxy::v1::config::StringMatch::kRegex:
+        // TODO: support regex
+        return false;
+        break;
+      default:
+        break;
```
Comment (lizan) [inline src/envoy/mixer/quota_config.cc]: Let's not allow this fall through and return `true`?

File: src/envoy/mixer/quota_config.cc (hunk 1 above)
Comment (lizan) [reply src/envoy/mixer/quota_config.cc]: ok fine.

File: src/envoy/mixer/quota_config.cc (hunk 4 above)
Comment (ayj) [reply src/envoy/mixer/quota_config.cc]: Should an `unknown` match type result in a valid match? Seems like the default behavior should be to ignore that case.

Comment (qiwzhang) [reply src/envoy/mixer/quota_config.cc]: @ayj I agreed with.  We should ignore "Unknown" type

Comment (lizan) [reply src/envoy/mixer/quota_config.cc]: No, there shouldn't be a unknown clause (it should be removed) and it shouldn't go through. You treated "unimplemented" regex as "no match" above, so the `unknown` should be same. Also leave a comment here.

Comment (qiwzhang) [reply src/envoy/mixer/quota_config.cc]: OK, I changed "unimplemented" as ignored.  so regex and not_set case are all ignored.

Comment (ayj) [reply src/envoy/mixer/quota_config.cc]: + 1 for comment explanation rationale. 

Really, this type of error should be caught earlier by validation before proxy receives the config. 

Shouldn't adding unoptimized regex support be fairly straightforward? e.g.

```c++
if (!std::regex_match(value, match.regex())) {
    return false
}
```

**Pull request merged**